File: Obj_Blocks_and_Swift_Closures/Objc_Blocks_Assignment/Objc_Blocks_Assignment/Question 1/CustomIntegerArray.m
//
//  CustomIntegerArray.m
//  Objc_Blocks_Assignment
//
//  Created by Mangesh Shiudkar on 26/08/26.
//

#import "CustomIntegerArray.h"

@implementation CustomIntegerArray {
    NSMutableArray<NSNumber *> *_storage;
}

- (instancetype)initWithValues:(NSArray<NSNumber *> *)values {

    self = [super init];
    
    if (self) {
        _storage = [values mutableCopy] ?: [NSMutableArray array];
    }
    
    return self;
}

- (NSArray<NSNumber *> *)allValues {
    return [_storage copy];
}

- (CustomIntegerArray *)filter:(FilterPredicateBlock)block {
    
    NSMutableArray<NSNumber *> *result = [NSMutableArray array];
    
    for (NSNumber *number in _storage) {
        
        NSInteger value = [number integerValue];
        
        if (block(value)) {
            [result addObject:number];
        }
    }
    
    return [[CustomIntegerArray alloc] initWithValues:result];
}

- (NSInteger)reduce:(ReduceBlock)block withInitial:(NSInteger)initialValue {
    
    NSInteger result = initialValue;
    
    for (NSNumber *number in _storage) {
        
        NSInteger value = [number integerValue];
        
        result = block(result, value);
    }
    
    return result;
}

- (CustomIntegerArray *)sortedArrayUsingBlock:(SortComparisonBlock)block {
    
    NSMutableArray<NSNumber *> *result = [_storage mutableCopy];
    
    NSUInteger count = result.count;
    
    for (NSUInteger i = 0; i < count; i++) {
        
        for (NSUInteger j = i + 1; j < count; j++) {
            
            NSInteger firstValue = [result[i] integerValue];
            NSInteger secondValue = [result[j] integerValue];
            
            NSComparisonResult comparison =
                block(firstValue, secondValue);
            
            if (comparison == NSOrderedDescending) {
                
                NSNumber *temp = result[i];
                result[i] = result[j];
                result[j] = temp;
            }
        }
    }
    
    return [[CustomIntegerArray alloc] initWithValues:result];
}

- (void)forEach:(ForEachBlock)block {
    
    for (NSNumber *number in _storage) {
        
        NSInteger value = [number integerValue];
        
        block(value);
    }
}

- (NSString *)description {
    return [_storage description];
}

@end

File: Obj_Blocks_and_Swift_Closures/Swift_Closures_Assignment/Swift_Closures_Assignment/Question 1/CustomIntegerArray.swift
//
//  CustomIntegerArray.swift
//  Swift_Closures_Assignment
//
//  Created by Mangesh Shiudkar on 26/08/26.
//


typealias FilterPredicateClosure = (Int) -> Bool

typealias ReduceClosure = (Int, Int) -> Int

typealias SortClosure = (Int, Int) -> Bool

typealias ForEachClosure = (Int) -> Void


class CustomIntegerArray {

    private var storage: [Int]

    init(values: [Int]) {
        storage = values
    }

    func filter(_ predicate: FilterPredicateClosure) -> CustomIntegerArray {
        var result: [Int] = []
        
        for value in storage {
            if predicate(value) {
                result.append(value)
            }
        }
        return CustomIntegerArray(values: result)
    }

    func reduce(_ operation: ReduceClosure, initialValue: Int) -> Int {
        var result = initialValue

        for value in storage {
            result = operation(result, value)
        }
        return result
    }

    func sorted(by comparison: SortClosure) -> CustomIntegerArray {
        var result = storage

        for i in 0..<result.count {
            for j in (i + 1)..<result.count {
                if comparison(result[j], result[i]) {
                    let temp = result[i]
                    result[i] = result[j]
                    result[j] = temp
                }
            }
        }
        return CustomIntegerArray(values: result)
    }

    func forEach(_ operation: ForEachClosure) {
        for value in storage {
            operation(value)
        }
    }
}


extension CustomIntegerArray: CustomStringConvertible {
    var description: String {
        return "\(storage)"
    }
}


func testCustomIntegerArray() {
    let numbers = CustomIntegerArray(
        values: [5, 2, 8, 1, 9, 3]
    )

    // 1. Filter
    let evens = numbers.filter { value in
        return value % 2 == 0
    }
    print("Evens: \(evens)")
    
    // 2. Reduce
    let sum = numbers.reduce(
        { partial, value in
            return partial + value
        },
        initialValue: 0
    )
    print("Sum: \(sum)")
    
    
    // 3. Sorted
    let sorted = numbers.sorted { a, b in
        return a < b
    }
    print("Sorted: \(sorted)")
    
    
    // 4. ForEach
    numbers.forEach { value in
        print("Value: \(value)")
    }
}
